src/geo/jts.clj

```diff
@@ -40,71 +143,206 @@
-  (multi-polygon (map polygon-wkt wkt)))
+  0 0]] [5 5 10 10 6 2]]. Allows an optional SRID argument at end."
+  ([wkt]
+   (multi-polygon (map polygon-wkt wkt)))
+  ([wkt srid]
+   (multi-polygon (map #(polygon-wkt % srid) wkt))))
+
+(defn- coord-dimension-check
+  "Return 3 if x, y, and z dimensions are valid. Return 2 if only x and y are valid."
+  [coord]
+  (let [x (not (Double/isNaN (.x coord)))
+        y (not (Double/isNaN (.y coord)))
+        z (not (Double/isNaN (.z coord)))]
+    (cond (and x y z)
+          3
+          (and x y)
+          2)))
+
+(defn coordinates
+  "Get a sequence of Coordinates from a Geometry"
+  [^Geometry geom]
+  (into [] (.getCoordinates geom)))
+
+(defn geom-dimension-check
+  "In a Geometry, return 3 if any Coordinate has a Z.
+  Otherwise return 2 if any Coordinate has a Y."
+  [geom]
+  (let [coords (coordinates geom)
+        dimensions (into [] (map coord-dimension-check coords))]
+    (cond (some #{3} dimensions)
+          3
+          (some #{2} dimensions)
+          2)))
+
+(defn same-srid?
+  "Check if two Geometries have the same SRID."
+  [^Geometry g1 ^Geometry g2]
+  (and (= (get-srid g1) (get-srid g2))
+       (not= (get-srid g1) 0)))
+
+(defn same-coords?
+  "Check if two Coordinates have the same number of dimensions and equal ordinates."
+  [^Coordinate c1 ^Coordinate c2]
+  (let [d1 (coord-dimension-check c1)
+        d2 (coord-dimension-check c2)]
+    (and (= d1 d2)
+         (cond (= d1 3)
+               (.equals3D c1 c2)
+               (= d1 2)
+               (.equals2D c1 c2)))))
+
+(defn same-geom?
+  "Check if two geometries are topologically equal, with the same SRID."
+  [^Geometry g1 ^Geometry g2]
+  (and (same-srid? g1 g2)
+       (.equalsTopo g1 g2)))
+
+(defn- transform-coord
+  "Transforms a coordinate using a proj4j transform.
+  Can either be specified with a transform argument or two projection arguments."
+  ([coord transform]
+   (-> (.transform transform
+                   (ProjCoordinate. (.x coord) (.y coord) (.z coord))
+                   (ProjCoordinate.))
+       (#(coordinate (.x %) (.y %) (.z %)))))
+  ([coord c1 c2]
+   (if (= c1 c2) coord
+                       (transform-coord coord (create-transform c1
+                                                                c2)))))
+
+(defn- transform-coord-seq-item
+  "Transforms one item in a CoordinateSequence using a proj4j transform."
+  [cseq i transform]
+  (let [coordinate (.getCoordinate cseq i)
+        transformed (transform-coord coordinate transform)]
+    (.setOrdinate cseq i 0 (.x transformed))
+    (.setOrdinate cseq i 1 (.y transformed))))
+
+(defn- transform-coord-seq-filter
+  "Implement JTS's CoordinateSequenceFilter, to be applied to a Geometry using transform-geom."
+  [transform]
+  (reify CoordinateSequenceFilter
+    (filter [_ seq i]
+      (transform-coord-seq-item seq i transform))
+    (isDone [_]
+      false)
+    (isGeometryChanged [_]
+      true)))
+
+(defn- tf-set-srid
```

Comment:
Again, my ignorance about these types of conversions and CRS's in general, but can you explain a bit more about how these 2 cases (SRID vs EPSG) are different, and whether it's really necessary to handle both?

I'd love to see some more explicit test cases that both verify these behavior and serve as documentation to explain how they might appear in normal usage.

Reply:
JTS geometries only take integers as their projection system. If .getSRID is 4326, that refers to the EPSG:4326 coordinate system.
proj4j (see [comment](https://github.com/Factual/geo/pull/10/files/0dee51ffd1fd263ed7794ae87afd758b2b82aeff#r193432759)) takes a wider view.

The thought process here is:
1. When the coordinate system (either source or target) is provided as an integer, assume it's an EPSG, convert it into a string called "EPSG:xxxx" and pass that along.
2. When the coordinate system (either source or target) is provided as a string, first try to parse it as a coordinate system name (whether EPSG or any of the other authority options), and failing that then try to parse it as a proj4 string.
3. After the transformation occurs, tf-set-srid comes into play: if the target coordinate system represents an EPSG projection -- either as "EPSG:4326" or "4326", then JTS is able to store that in its srid metadata. If that's the case, do so during a projection.
4. When a target coordinate system represents anything else -- whether a string name that isn't "EPSG:" or proj4, then the transformation stills occur to the coordinates themselves, but nothing should be set for the JTS srid.
5. When the source and target projection systems are identical, clone the object without passing it through a transform. When that projection is an EPSG code, set the JTS's SRID to that.

Separately, there is an extraneous -> macro in transform-geom I also need to remove.